src/sage/rings/polynomial/polydict.pyx

```diff
@@ -200,13 +206,53 @@ cdef class PolyDict:
             sage: f.remove_zeros()
             sage: f
             PolyDict with representation {}
+
+            sage: R.<t> = PowerSeriesRing(QQ)
+            sage: f = PolyDict({(1, 1): O(t), (1, 0): R.zero()})
+            sage: f.remove_zeros(lambda s: s.is_zero() and s.prec() is Infinity)
+            sage: f
+            PolyDict with representation {(1, 1): O(t^1)}
         """
-        if not all(self.__repn.values()):
+        if not self.__repn:
+            return
+        cdef bint has_zero_coefficient = False
+        if zero_test is None:
+            for coeff in self.__repn.values():
+                if not coeff:
+                    has_zero_coefficient = True
+                    break
+            if not has_zero_coefficient:
+                return
```

Comment:
Rather than going through everything potentially twice, I would either get a list of keys or do something like
```python
if all(self.__repn.values()):
    return
# otherwise there is at least one 0, so filter it out
```
Similar comment for the `zero_function` case since you had said such generator checks are faster.

Reply:
Comment added in https://github.com/sagemath/sage/pull/35174/commits/1a7eeac1a5da609b30fa9f0a36d6784e5dff6e7d